### ISSUE TITLE
Resolve user variables when resolving other user variables (user var composition)

### ISSUE THREAD

**Issue by fschmager** [mitchellh/packer#3161]
**tl;dr** Allow composition of user variables using other user variables.

**Extended**
In the example a user variable `ami_name` is defined. It should be suffixed with the value or another user variable `name`. `name` defaults to "default". User may override via `-var` parameter. 
However, packer does not seem to be using the value of `name` in the the other user variable `ami_name`. 

```javascript
{
  "variables":{
    "name":"default",
    "ami_name":"ami-{{user `name`}}"
  },
  "builders":[
    {
      "type":"amazon-ebs",
      "ami_name":"{{user `ami_name`}}",
      ...
    }
  ],
  ...
}
```

**Example1:**
```$ packer build -var 'name=bar' template.json``` 

_Expected:_
Creates an AMI with the name "ami-bar"

_Actual:_
Creates an AMI with the name "ami-"

**Example2:**
```$ packer build template.json``` 

_Expected:_
Creates an AMI with the name "ami-default"

_Actual:_
Creates an AMI with the name "ami-"

**Log snippet**
The log snippet blow shows that the value of `ami_name` is just "ami-".
```
...
==> amazon-ebs: Waiting for the instance to stop...
==> amazon-ebs: Creating the AMI: ami-
    amazon-ebs: AMI: ami-12345678
==> amazon-ebs: Waiting for AMI to become ready...
...
```

It seems that packer templating does not resolve user variables within the `variables` section.

Please support the composition of user variables. If I wanted to use the value of `ami_name` as the value for the actual ami_name AND as the value in an ami tag, I would need to do the composition in two places. And we know, duplication is the root of all evil.

```
$ packer --version
0.8.6
```

Complete [template and log gist](https://gist.github.com/fschmager/4792856a56fe1e80c862).

**Comment by cbednarski** [comment]
Thanks for opening an issue. Variable interpolation is pretty complicated as it is and this would make it even more complicated because we'd have to do it recursively. I understand the use-case but I don't want to take a complexity hit in this part of packer. I think you can achieve approximately the same thing using something like this:

```
{
  "variables":{
    "name":"default"
  },
  "builders":[
    {
      "type":"amazon-ebs",
      "ami_name":"ami-{{user `name`}}",
      ...
    }
  ],
  ...
}
```

There are also other ways to do this outside of packer, for instance using environment variables or using a template pre-processor like [racker](https://github.com/aspring/racker).
